src/tsolvers/egraph/EgraphSolver.cc

```diff
@@ -292,94 +267,32 @@ void Egraph::declareTermRecursively(PTRef tr) {
-        }
-
-        // Add the pure term
-        ERef er_pure = enode_store.addTerm(sym, ERef_Nil, tr_pure);
-        // Add the negated term
-        ERef er_neg = enode_store.addTerm(enode_store.sym_uf_not, enode_store.addList(er_pure, ERef_Nil), logic.mkNot(tr_pure));
-
-        updateParentsVector(tr_pure);
-        updateParentsVector(logic.mkNot(tr_pure));
+void Egraph::declareTerm(PTRef tr) {
+    if (!isValid(tr) && !logic.isTheoryTerm(tr) && !logic.isBoolAtom(tr)) { return; }
+    if ((logic.isBoolAtom(tr) || logic.isBooleanOperator(tr)) && !logic.appearsInUF(tr)) { return; }
 
-        // Make sure er_pure and er_neg need to be different
-        assertNEq(er_pure, er_neg, PtAsgn_Undef);
 
-        boolTermToERef.insert(tr_pure, er_pure);
-        boolTermToERef.insert(tr_neg, er_neg);
-    } else {
+    auto PTRefERefPairVec = enode_store.constructTerm(tr);
 
-        if (not logic.isIte(tr)) {
-            sym = enode_store.addSymb(tm.symb());
-            cdr = ERef_Nil;
-            for (int j = tm.size() - 1; j >= 0; j--) {
-                assert(enode_store.termToERef.has(tm[j])); // The child was not inserted
-                ERef car = enode_store.termToERef[tm[j]];
 #ifdef VERBOSE_EUF
-                ERef prev_cdr = cdr;
-                assert (enode_store[car].getRoot() == car);
-                assert (enode_store[cdr].getRoot() == cdr);
+    cerr << "EgraphSolver: Adding term " << logic.printTerm(tr) << " (" << tr.x << ")" << endl;
 #endif
-                cdr = enode_store.addList(car, cdr);
-            }
-        } else {
-            sym = enode_store.addIteSymb(tr);
-            cdr = ERef_Nil;
-        }
-        ERef er = enode_store.addTerm(sym, cdr, tr);
 
-        updateParentsVector(tr);
-
-        if (logic.isUP(tr) || logic.isEquality(tr)) {
-            PTRef tr_neg = logic.mkNot(tr);
-            ERef er_neg = enode_store.addTerm(enode_store.sym_uf_not, enode_store.addList(er, ERef_Nil), tr_neg);
-            updateParentsVector(tr_neg);
-            boolTermToERef.insert(tr, er);
-            boolTermToERef.insert(tr_neg, er_neg);
-        }
+    if (PTRefERefPairVec.size() == 0) {
+        return;
     }
-}
 
-//
-// No recursion here, we assume the caller has already introduced the
-// subterms
-//
-void Egraph::declareTerm(PTRef tr) {
-
-    if (!isValid(tr) && !logic.isTheoryTerm(tr) && !logic.isBoolAtom(tr)) { return; }
-    if ((logic.isBoolAtom(tr) || logic.isBooleanOperator(tr)) && !logic.appearsInUF(tr)) { return; }
-    constructTerm(tr);
-#ifdef VERBOSE_EUF
-        cerr << "EgraphSolver: Adding term " << logic.printTerm(tr) << " (" << tr.x << ")" << endl;
-#endif
+    for (auto PTRefERefPair : PTRefERefPairVec) {
+        updateParentsVector(PTRefERefPair.first);
+    }
 
-    if (logic.isDisequality(tr) && !enode_store.dist_classes.has(tr))
-        enode_store.addDistClass(tr);
+    if (logic.hasSortBool(tr) && !logic.isDistinct(tr)) {
```

Comment:
I think we have established that boolean distincts are always simplified, so the second check seems redundant, no?

Reply:
Redundant check removed.

Reply:
Scratch that. I confused the sort of the arguments and the return sort. This check is not redundant, all distincts (regardless of the sort of the arguments) have return sort boolean, which is what the first part of the condition checks.